diff --git a/ChangeLog b/ChangeLog
index e9bc955..2a7a5d5 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -16,6 +16,13 @@
      
      Reported by mcv21 AT cam.ac.uk
 
+20131229
+ - (djm) OpenBSD CVS Sync
+   - djm@cvs.openbsd.org 2013/12/19 00:10:30
+     [ssh-add.c]
+     skip requesting smartcard PIN when removing keys from agent; bz#2187
+     patch from jay AT slushpupie.com; ok dtucker
+
 20131109
  - (dtucker) [configure.ac kex.c key.c myproposal.h] Test for the presence of
    NID_X9_62_prime256v1, NID_secp384r1 and NID_secp521r1 and test that the
diff --git a/ssh-add.c b/ssh-add.c
index d268cdd..a8144c8 100644
--- a/ssh-add.c
+++ b/ssh-add.c
@@ -257,14 +257,17 @@ add_file(AuthenticationConnection *ac, const char *filename)
 static int
 update_card(AuthenticationConnection *ac, int add, const char *id)
 {
-	char *pin;
+	char *pin = NULL;
 	int ret = -1;
 
-	pin = read_passphrase("Enter passphrase for PKCS#11: ", RP_ALLOW_STDIN);
-	if (pin == NULL)
-		return -1;
+	if (add) {
+		if ((pin = read_passphrase("Enter passphrase for PKCS#11: ",
+		    RP_ALLOW_STDIN)) == NULL)
+			return -1;
+	}
 
-	if (ssh_update_card(ac, add, id, pin, lifetime, confirm)) {
+	if (ssh_update_card(ac, add, id, pin == NULL ? "" : pin,
+	    lifetime, confirm)) {
 		fprintf(stderr, "Card %s: %s\n",
 		    add ? "added" : "removed", id);
 		ret = 0;
@@ -273,7 +276,8 @@ update_card(AuthenticationConnection *ac, int add, const char *id)
 		    add ? "add" : "remove", id);
 		ret = -1;
 	}
-	xfree(pin);
+	if (pin)
+		xfree(pin);
 	return ret;
 }
 
