diff -up openssh-4.3p2/contrib/ssh-copy-id.identity openssh-4.3p2/contrib/ssh-copy-id
--- openssh-4.3p2/contrib/ssh-copy-id.identity	2011-10-19 10:40:09.611423751 +0200
+++ openssh-4.3p2/contrib/ssh-copy-id	2011-10-19 10:40:54.362460032 +0200
@@ -1,11 +1,11 @@
 #!/bin/sh
 
-# Shell script to install your identity.pub on a remote machine
+# Shell script to install your id_rsa.pub on a remote machine
 # Takes the remote machine name as an argument.
 # Obviously, the remote machine must accept password authentication,
 # or one of the other keys in your ssh-agent, for this to work.
 
-ID_FILE="${HOME}/.ssh/identity.pub"
+ID_FILE="${HOME}/.ssh/id_rsa.pub"
 
 if [ "-i" = "$1" ]; then
   shift
