diff -up openssh-6.1p1/openbsd-compat/port-linux.c.privsep-selinux openssh-6.1p1/openbsd-compat/port-linux.c
--- openssh-6.1p1/openbsd-compat/port-linux.c.privsep-selinux	2012-10-12 13:35:03.715980297 +0200
+++ openssh-6.1p1/openbsd-compat/port-linux.c	2012-10-12 13:35:03.719980279 +0200
@@ -505,6 +505,23 @@ ssh_selinux_change_context(const char *n
 	xfree(newctx);
 }
 
+void
+ssh_selinux_copy_context(void)
+{
+	char *ctx;
+
+	if (!ssh_selinux_enabled())
+		return;
+
+	if (getexeccon((security_context_t *)&ctx) != 0) {
+		logit("%s: getcon failed with %s", __func__, strerror (errno));
+		return;
+	}
+	if (setcon(ctx) != 0)
+		logit("%s: setcon failed with %s", __func__, strerror (errno));
+	xfree(ctx);
+}
+
 #endif /* WITH_SELINUX */
 
 #ifdef LINUX_OOM_ADJUST
diff -up openssh-6.1p1/openbsd-compat/port-linux.h.privsep-selinux openssh-6.1p1/openbsd-compat/port-linux.h
--- openssh-6.1p1/openbsd-compat/port-linux.h.privsep-selinux	2011-01-25 02:16:18.000000000 +0100
+++ openssh-6.1p1/openbsd-compat/port-linux.h	2012-10-12 13:35:03.719980279 +0200
@@ -24,6 +24,7 @@ int ssh_selinux_enabled(void);
 void ssh_selinux_setup_pty(char *, const char *);
 void ssh_selinux_setup_exec_context(char *);
 void ssh_selinux_change_context(const char *);
+void ssh_selinux_copy_context(void);
 void ssh_selinux_setfscreatecon(const char *);
 #endif
 
diff -up openssh-6.1p1/session.c.privsep-selinux openssh-6.1p1/session.c
--- openssh-6.1p1/session.c.privsep-selinux	2012-10-12 13:35:03.670980503 +0200
+++ openssh-6.1p1/session.c	2012-10-12 14:03:01.011305806 +0200
@@ -1513,6 +1513,10 @@ do_setusercontext(struct passwd *pw)
 
 		platform_setusercontext_post_groups(pw);
 
+
+#ifdef WITH_SELINUX
+		ssh_selinux_copy_context();
+#endif
 		if (options.chroot_directory != NULL &&
 		    strcasecmp(options.chroot_directory, "none") != 0) {
                         tmp = tilde_expand_filename(options.chroot_directory,
@@ -1787,9 +1791,6 @@ do_child(Session *s, const char *command
 		argv[i] = NULL;
 		optind = optreset = 1;
 		__progname = argv[0];
-#ifdef WITH_SELINUX
-		ssh_selinux_change_context("sftpd_t");
-#endif
 		exit(sftp_server_main(i, argv, s->pw));
 	}
 
diff -up openssh-6.1p1/sshd.c.privsep-selinux openssh-6.1p1/sshd.c
--- openssh-6.1p1/sshd.c.privsep-selinux	2012-10-12 13:35:03.716980292 +0200
+++ openssh-6.1p1/sshd.c	2012-10-12 13:35:03.721980271 +0200
@@ -794,6 +794,13 @@ privsep_postauth(Authctxt *authctxt)
 	do_setusercontext(authctxt->pw);
 
  skip:
+#ifdef WITH_SELINUX
+	/* switch SELinux content for root too */
+	if (authctxt->pw->pw_uid == 0) {
+		ssh_selinux_copy_context();
+	}
+#endif
+
 	/* It is safe now to apply the key state */
 	monitor_apply_keystate(pmonitor);
 
