diff --git a/session.c b/session.c
index 8b6db20..0bdf638 100644
--- a/session.c
+++ b/session.c
@@ -1805,15 +1805,6 @@ do_child(Session *s, const char *command)
 			exit(1);
 	}
 
-	/*
-	 * Close the connection descriptors; note that this is the child, and
-	 * the server will still have the socket open, and it is important
-	 * that we do not shutdown it.  Note that the descriptors cannot be
-	 * closed before building the environment, as we call
-	 * get_remote_ipaddr there.
-	 */
-	child_close_fds();
-
 	if (!options.use_login)
 		do_rc_files(s, shell);
 
@@ -1849,6 +1840,15 @@ do_child(Session *s, const char *command)
 		exit(sftp_server_main(i, argv, s->pw));
 	}
 
+	/*
+	 * Close the connection descriptors; note that this is the child, and
+	 * the server will still have the socket open, and it is important
+	 * that we do not shutdown it.  Note that the descriptors cannot be
+	 * closed before building the environment, as we call
+	 * get_remote_ipaddr there.
+	 */
+	child_close_fds();
+
 	if (options.use_login) {
 		launch_login(pw, hostname);
 		/* NEVERREACHED */
