diff --git a/ChangeLog b/ChangeLog
index 45e6942..e9bc955 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -22,6 +22,14 @@
    latter actually works before using it.  Fedora (at least) has NID_secp521r1
    that doesn't work (see https://bugzilla.redhat.com/show_bug.cgi?id=1021897).
 
+20131023
+ - (djm) OpenBSD CVS Sync
+   - djm@cvs.openbsd.org 2013/10/23 04:16:22
+     [ssh-keygen.c]
+     Make code match documentation: relative-specified certificate expiry time
+     should be relative to current time and not the validity start time.
+     Reported by Petr Lautrbach; ok deraadt@
+
 20130922
  - (dtucker) [platform.c platform.h sshd.c] bz#2156: restore Linux oom_adj
    setting when handling SIGHUP to maintain behaviour over retart.  Patch
diff --git a/ssh-keygen.c b/ssh-keygen.c
index c3dad3b..25d8613 100644
--- a/ssh-keygen.c
+++ b/ssh-keygen.c
@@ -1385,7 +1385,7 @@ parse_cert_times(char *timespec)
 		cert_valid_from = parse_absolute_time(from);
 
 	if (*to == '-' || *to == '+')
-		cert_valid_to = parse_relative_time(to, cert_valid_from);
+		cert_valid_to = parse_relative_time(to, now);
 	else
 		cert_valid_to = parse_absolute_time(to);
 
