diff --git a/ChangeLog b/ChangeLog
index 3304206..45e6942 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -22,6 +22,11 @@
    latter actually works before using it.  Fedora (at least) has NID_secp521r1
    that doesn't work (see https://bugzilla.redhat.com/show_bug.cgi?id=1021897).
 
+20130922
+ - (dtucker) [platform.c platform.h sshd.c] bz#2156: restore Linux oom_adj
+   setting when handling SIGHUP to maintain behaviour over retart.  Patch
+   from Matthew Ife.
+
 20111104
  - (dtucker) OpenBSD CVS Sync
    - djm@cvs.openbsd.org 2011/10/24 02:10:46
diff --git a/platform.c b/platform.c
index 0b02e81..d6b28f6 100644
--- a/platform.c
+++ b/platform.c
@@ -39,6 +39,14 @@ platform_pre_fork(void)
 }
 
 void
+platform_pre_restart(void)
+{
+#ifdef LINUX_OOM_ADJUST
+	oom_adjust_restore();
+#endif
+}
+
+void
 platform_post_fork_parent(pid_t child_pid)
 {
 #ifdef USE_SOLARIS_PROCESS_CONTRACTS
diff --git a/platform.h b/platform.h
index 5e65cff..915b70b 100644
--- a/platform.h
+++ b/platform.h
@@ -20,5 +20,6 @@
 
 void platform_pre_listen(void);
 void platform_pre_fork(void);
+void platform_pre_restart(void);
 void platform_post_fork_parent(pid_t child_pid);
 void platform_post_fork_child(void);
diff --git a/sshd.c b/sshd.c
index a27a082..41835e3 100644
--- a/sshd.c
+++ b/sshd.c
@@ -323,6 +323,7 @@ static void
 sighup_restart(void)
 {
 	logit("Received SIGHUP; restarting.");
+	platform_pre_restart();
 	signal(SIGHUP, SIG_IGN); /* will be restored after exec */
 	close_listen_socks();
 	close_startup_pipes();
