diff --git a/PROTOCOL b/PROTOCOL
index 131adfe..c828087 100644
--- a/PROTOCOL
+++ b/PROTOCOL
@@ -328,6 +328,11 @@ a server may offer multiple keys of the same type for a period (to
 give clients an opportunity to learn them using this extension) before
 removing the deprecated key from those offered.
 
+2.6. connection: add IUTF8 terminal mode flag
+
+OpenSSH supports the IUTF8 terminal mode flag and encodes it in "pty-req"
+messages as opcode value 42.
+
 3. SFTP protocol changes
 
 3.1. sftp: Reversal of arguments to SSH_FXP_SYMLINK
diff --git a/ttymodes.h b/ttymodes.h
index 4d848fe..396ae88 100644
--- a/ttymodes.h
+++ b/ttymodes.h
@@ -127,6 +127,9 @@ TTYMODE(IXOFF,	c_iflag, 40)
 #ifdef IMAXBEL
 TTYMODE(IMAXBEL,c_iflag, 41)
 #endif /* IMAXBEL */
+#ifdef IUTF8
+TTYMODE(IUTF8,	c_iflag, 42)
+#endif /* IUTF8 */
 
 TTYMODE(ISIG,	c_lflag, 50)
 TTYMODE(ICANON,	c_lflag, 51)

